File: code/backend/Odin/src/main/kotlin/pt/isel/odin/service/interfaces/StudentService.kt
package pt.isel.odin.service.interfaces

import pt.isel.odin.controller.dto.student.StudentRequest
import pt.isel.odin.model.Student

/**
 * Service for students.
 */
interface StudentService {

    /**
     * Gets a student by its id.
     *
     * @param id the student id.
     */
    fun getById(id: String): Student

    /**
     * Gets a student by its email.
     *
     * @param email the student email.
     */
    fun getByEmail(email: String): Student?

    /**
     * Gets all students.
     */
    fun getAll(): List<Student>

    /**
     * Saves or Updates a student.
     *
     * @param studentRequest the student to save.
     */
    fun save(studentRequest: StudentRequest): Student

    /**
     * Deletes a student by its id.
     *
     * @param id the student id.
     */
    fun delete(id: String)
}

File: code/frontend/Tyr/src/services/StudentService.ts
import {makeApiRequest} from '../axios/apiRequest';
import {Student} from "../model/Student";

/**
 * Service to manage students
 */
export class StudentService {

    static basePath = '/students';

    static async getById(id: number): Promise<Student> {
        return makeApiRequest('get', `${(StudentService.basePath)}/${id}`);
    }

    static async getAll(): Promise<Student[]> {
        return makeApiRequest('get', StudentService.basePath);
    }

    static async save(studentRequest: Student): Promise<Student> {
        return makeApiRequest('post', `${(StudentService.basePath)}/save`, studentRequest);
    }

    static async update(studentRequest: Student): Promise<Student> {
        return makeApiRequest('put', `${(StudentService.basePath)}/update`, studentRequest);
    }

    static async delete(id: number): Promise<void> {
        return makeApiRequest('delete', `${(StudentService.basePath)}/${id}`);
    }
}